utils/utils.c

```diff
@@ -845,3 +849,64 @@ int AlmostEqualRelativeAndAbs_double(double A, double B,
 }
 
 /* #undef __USE_XOPEN2K */
+
+/* A parallel cumulative sum
+   Output convention is: cumsum[0] = 0; cumsum[N-1] = sum(a[0:N-1]);
+   The algorithm is:
+   - Divide the array into `nthreads` chunks
+   - cumsum within each chunk
+   - compute the "offset" for each chunk by summing the cumsum at the tail of all previous chunks
+   - apply the offset
+*/
+void parallel_cumsum(const int64_t *a, const int64_t N, int64_t *cumsum){
+    #ifdef _OPENMP
+    int nthreads = omp_get_max_threads();
+    #else
+    int nthreads = 1;
+    #endif
+    
+    int64_t min_N_per_thread = 10000;  // heuristic to avoid overheads
+    if(N/min_N_per_thread < nthreads){
+        nthreads = N/min_N_per_thread;
+    }
+    if(nthreads < 1){
+        nthreads = 1;
+    }
```

Comment:
Looks like you are changing the number of threads used for the computation, upper-limited by `max_threads`. But `max_threads` could be quite large (say 32) while requested `nthreads` could be 4. Is it possible for the code to use a larger number of threads than requested?

Reply:
`OMP_NUM_THREADS` and `omp_set_num_threads()` control `omp_get_max_threads()`, so this should be safe.